File: 212-word-search-ii/word-search-ii.java
class TrieNode {
    Map<Character, TrieNode> children;
    boolean isEndOfWord;

    public TrieNode() {
        children = new HashMap<>();
        isEndOfWord = false;
    }
}

class Trie {
    TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode node = root;
        for (char ch : word.toCharArray()) {
            node.children.putIfAbsent(ch, new TrieNode());
            node = node.children.get(ch);
        }
        node.isEndOfWord = true;
    }
}

class Solution {
    public List<String> findWords(char[][] board, String[] words) {
        List<String> result = new ArrayList<>();
        Trie trie = new Trie();

        for (String word : words) {
            trie.insert(word);
        }

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                char startLetter = board[i][j];
                if (trie.root.children.containsKey(startLetter)) {
                    backtrack(board, i, j, trie.root.children.get(startLetter), result, "" + startLetter);
                }
            }
        }

        return result;
    }

    private void backtrack(char[][] board, int i, int j, TrieNode trieNode, List<String> result, String path) {
    if (trieNode.isEndOfWord) {
        result.add(path);
        trieNode.isEndOfWord = false;
    }

    char original = board[i][j];
    board[i][j] = '#';

    int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    for (int[] dir : directions) {
        int ni = i + dir[0];
        int nj = j + dir[1];
        if (ni >= 0 && ni < board.length && nj >= 0 && nj < board[0].length && board[ni][nj] != '#' && trieNode.children.containsKey(board[ni][nj])) {
            backtrack(board, ni, nj, trieNode.children.get(board[ni][nj]), result, path + board[ni][nj]);
        }
    }

    board[i][j] = original;
}
}

File: 212-word-search-ii/word-search-ii.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEndOfWord = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.isEndOfWord = True

        
class Solution(object):
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        result = []
        trie = Trie()

        for word in words:
            trie.insert(word)

        def backtrack(node, i, j, path):
            if node.isEndOfWord:
                result.append(path)
                node.isEndOfWord = False

            original = board[i][j]
            board[i][j] = '#'

            directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
            for di, dj in directions:
                ni, nj = i + di, j + dj
                if 0 <= ni < len(board) and 0 <= nj < len(board[0]) and board[ni][nj] != '#' and board[ni][nj] in node.children:
                    backtrack(node.children[board[ni][nj]], ni, nj, path + board[ni][nj])

            board[i][j] = original

        for i in range(len(board)):
            for j in range(len(board[0])):
                start_letter = board[i][j]
                if start_letter in trie.root.children:
                    backtrack(trie.root.children[start_letter], i, j, start_letter)

        return result